### PR TITLE
[draft]Fix the tensor save for backward release problem

### PR DESCRIPTION
**Pull request by guyang3532**
Motivation:
PythonOp is saving input for backward, it's risky since ONNX Runtime backend is not aware of this, the tensor buffer may be "released" by ORT, then potentially modified by other operators before backward function executes.

Fix:
This pr just add a check that if input of PythonOp is saved for backward, and if input saved for backward is cloned for save.
If not, just throw an RuntimeError that alert user to clone the input before saved for backward like changing 'ctx.save_for_backward(input)' to 'ctx.save_for_backward(input.detach().clone())'

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py [hunk 1]
```diff
@@ -164,6 +164,13 @@ def register_context(result):
 
             # Extract results as DLPack tensors plus autograd context. Also skips all None values.
             unwrapped_values = wrap_all_outputs(result, is_training_mode)
+            ctx = unwrapped_values[0]
+            if ctx and ctx.saved_tensors:
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:168]: Shall we restrict the check for "is_training_mode=True"?

Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:168]: we can unpack ctx.saved_tensors just once and store as local variable.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py [hunk 2]
```diff
@@ -164,6 +164,13 @@ def register_context(result):
...
+                for tensor in ctx.saved_tensors:
+                    if tensor in set(new_wrapped_args):
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:170]: tensor could be None, better to add a check here.

Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:170]: converted to set just once outside the loop?

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py [hunk 3]
```diff
@@ -164,6 +164,13 @@ def register_context(result):
...
+                        raise wrap_exception(
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:171]: I think you can just raise RuntimeError() here, and it will be caught in the line 169 and wrapped as another exception.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py [hunk 4]
```diff
@@ -164,6 +164,13 @@ def register_context(result):
...
+                            RuntimeError(f"When save input for backward, you need to clone and save its copy not itself.")
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:172]: Remove f if there is no variable to output as the pylint complains. 

Shall we give a more concrete info how to mitigate the issue? 

"PythonOp is saving input for backward, it's risky since ONNX Runtime backend is not aware of this, the tensor  buffer may be "released" by ORT, then potentially modified by other operators before backward function executes.
Try to mitigate this risk by changing 'ctx.save_for_backward(input)' to 'ctx.save_for_backward(input.detach().clone())'."

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py [hunk 5]
```diff
@@ -1272,3 +1272,91 @@ def forward(self, model_input):
     ortmodule = ORTModule(TestModel(output_size)).eval()
     _ = ortmodule(torch.randn(output_size, dtype=torch.float))
     check_pythonop_training_mode(ortmodule, is_eval_mode=True)
+
+
+def test_python_op_input():
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:1277]: -> test_python_op_save_input_for_backward

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py [hunk 6]
```diff
@@ -1272,3 +1272,91 @@ def forward(self, model_input):
...
+            return activation4
+
+    class TestModule(torch.nn.Module):
+        def __init__(self, output_size, input_copy=False) -> None:
+            super().__init__()
+            self.layers = torch.nn.ModuleList([TestLayer(output_size, input_copy) for i in range(10)])
+
+        def forward(self, x):
+            # ModuleList can act as an iterable, or be indexed using ints
+            for i, l in enumerate(self.layers):
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:1340]: nit: better to fix those pylint warnings. :)

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py [hunk 7]
```diff
@@ -1272,3 +1272,91 @@ def forward(self, model_input):
...
+            g = ff * grad_output
+            return g
+
+    class TestLayer(torch.nn.Module):
+        def __init__(self, output_size, input_copy=False):
+            super().__init__()
+            if input_copy:
+                self.relu = GeLUFunctionTakeActivationInputCopy.apply
+            else:
+                self.relu = GeLUFunctionTakeActivationInput.apply
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:1316]: nit: maybe we can pass 'input_copy' as input of GeLUFunctionTakeActivationInput.apply?

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py [hunk 8]
```diff
@@ -1272,3 +1272,91 @@ def forward(self, model_input):
...
+
+    assert not can_run
+
+    def model_builder():
+        return TestModule(output_size, True)
+    def input_generator():
+        return torch.randn(output_size, output_size , dtype=torch.float).requires_grad_()
+    label_input = torch.ones([output_size])
+    run_training_test_and_compare(model_builder, input_generator, label_input)
+
```
Comment (pengwa) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:1362]: Looking at other tests in this file, seems we are trying to save input for backward at many places. :) Maybe we should add copy for them all to make them pass.

**Review by pengwa — COMMENTED** [review]
Thanks a lot for the temporarily fix. Leave few comments.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (hunk 1 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: thx, done.

Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: ? but ctx.saved_tensors is only used once

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (hunk 2 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: thx, done

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (hunk 3 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: thx, done.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (hunk 4 above)
Comment (guyang3532) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: thx, done.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py (hunk 5 above)
Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py]: thx, done.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py (hunk 6 above)
Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py]: ok

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py (hunk 8 above)
Comment (guyang3532) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py]: thx, done.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py (hunk 7 above)
Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_autograd.py:1316]: I mean we only have GeLUFunctionTakeActivationInput class, and in the forward function, there is a argument saying whether it did copy or not when save for backward.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py]: line 168, and line 169 are both triggering the unpacking operation.

File: orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py [hunk 9]
```diff
@@ -164,6 +164,17 @@ def register_context(result):
 
             # Extract results as DLPack tensors plus autograd context. Also skips all None values.
             unwrapped_values = wrap_all_outputs(result, is_training_mode)
+            ctx = unwrapped_values[0]
+            if is_training_mode and ctx and ctx.saved_tensors:
+                new_wrapped_args_set = set(new_wrapped_args)
+                for tensor in ctx.saved_tensors:
+                    if tensor is not None and tensor in new_wrapped_args_set:
+                        raise RuntimeError("PythonOp is saving input for backward, \
```
Comment (pengwa) [inline orttraining/orttraining/python/training/ortmodule/_custom_autograd_function_runner.py:172]: nit: let's add a TODO, we need a follow up fix instead of throwing the error.
